modules/core/02-client/keeper/extended_keeper.go

```diff
@@ -57,242 +52,80 @@ func (ek *ExtendedKeeper) SetHooks(ch ClientHooks) *ExtendedKeeper {
-			return err
-		}
-
-		// Upon other errors (that happen under honest majority), return error without timestamping it
-		if err != nil {
-			return err
-		}
-
-		// the header has a valid QC and is extending canonical chain
-		// follow the original verification rules to update client state
-		return ek.Keeper.UpdateClient(ctx, clientID, header)
-	case *ibctmtypes.Misbehaviour:
-		// on Misbehaviour,just forward to standard client logic
-		return ek.Keeper.UpdateClient(ctx, clientID, header)
-	default:
-		return types.ErrInvalidClientType
-	}
-}
-
-// checkHeader checks
-// - if a header is on fork,
-// - if a header has a valid QC or not, and
-// - other miscellaneous verification rules
-// It is essentially `CheckHeaderAndUpdateState` without updating the state
-// (adapted from https://github.com/cosmos/ibc-go/blob/v5.0.0/modules/core/02-client/keeper/client.go#L60)
-func (ek ExtendedKeeper) checkHeader(ctx sdk.Context, clientID string, tmHeader *ibctmtypes.Header) error {
+func (ek ExtendedKeeper) UpdateClient(ctx sdk.Context, clientID string, clientMsg exported.ClientMessage) error {
 	clientState, found := ek.GetClientState(ctx, clientID)
 	if !found {
 		return sdkerrors.Wrapf(types.ErrClientNotFound, "cannot update client with ID %s", clientID)
 	}
 
 	clientStore := ek.ClientStore(ctx, clientID)
+
 	if status := clientState.Status(ctx, clientStore, ek.cdc); status != exported.Active {
 		return sdkerrors.Wrapf(types.ErrClientNotActive, "cannot update client (%s) with status %s", clientID, status)
 	}
 
-	// Check if the header is on a fork
-	// If the consensus state exists, and it does not match this header, then this header is on a fork
-	// (adapted from https://github.com/cosmos/ibc-go/blob/v5.0.0/modules/light-clients/07-tendermint/types/update.go#L63-L77)
-	isOnFork := false
-	prevConsState, _ := ibctmtypes.GetConsensusState(clientStore, ek.cdc, tmHeader.GetHeight())
-	if prevConsState != nil {
-		if !reflect.DeepEqual(prevConsState, tmHeader.ConsensusState()) {
-			isOnFork = true
-		}
-	}
-
-	// Check if the header can pass all verification rules, including the QC
-	// Note that the first header on a fork can also be valid
-	// (adapted from https://github.com/cosmos/ibc-go/blob/v5.0.0/modules/light-clients/07-tendermint/types/update.go#L79-L89)
-	trustedConsState, exists := ibctmtypes.GetConsensusState(clientStore, ek.cdc, tmHeader.TrustedHeight)
-	if !exists {
-		return sdkerrors.Wrapf(
-			types.ErrConsensusStateNotFound, "could not get consensus state from clientstore at TrustedHeight: %s", tmHeader.TrustedHeight,
-		)
-	}
-	// asserting clientState to that of Tendermint client
-	cs, ok := clientState.(*ibctmtypes.ClientState)
-	if !ok {
-		return sdkerrors.Wrapf(
-			types.ErrFailedClientStateVerification, "expected type %T, got %T", &ibctmtypes.ClientState{}, clientState,
-		)
-	}
-	if err := checkValidity(cs, trustedConsState, tmHeader, ctx.BlockTime()); err != nil {
```

Comment:
Where does these checks go in this PR? They are kind of an integral part for Babylon and behave differently from the original IBC-Go implementation. Specifically, here we intercept a header that have a valid commit but are conflicted with another header. Without intercepting such headers the light client will freeze directly upon them.

Reply:
I see, thanks for the explanation. This definitely makes our life much easier! I wish I was developing our fork based on v6 rather than v5 😓

Reply:
yea, I think this refactor will be available only form v7 due to this refactor - https://medium.com/the-interchain-foundation/client-refactor-laying-the-groundwork-for-ibc-to-expand-across-ecosystems-61ec5a1b63bc